docs/isolated-cpu-affinity-transition.md

```diff
@@ -0,0 +1,74 @@
+## Isolated CPU affinity transition
+
+The introduction of the kernel commit 46a87b3851f0d6eb05e6d83d5c5a30df0eca8f76
+has affected a deterministic scheduling behavior by distributing tasks
+across CPU cores within a cgroups cpuset. It means that some runc operations
+like `runc exec` might be impacted under some circumstances, by example when
+a container has been created within a cgroup cpuset entirely composed of
+isolated CPU cores usually sets either with `nohz_full` and/or `isolcpus`
+kernel boot parameters.
+
+Some containerized real-time applications are relying on this deterministic
+behavior and uses the first CPU core to run a slow thread while other CPU
+cores are fully used by the real-time threads with SCHED_FIFO policy.
+Such applications can prevent runc process from joining a container when the
+runc process is randomly scheduled on a CPU core owned by a real-time thread.
+
+Runc introduces a way to restore this behavior via a dedicated kernel boot
+parameter:
+
+`runc.exec.isolated-cpu-affinity-transition`
```

Comment:
Why not use:
- an annotation in config.json
- a CLI flag for runc
- env var

Reply:
This PR essentially targets kubernetes environments using realtime tuned profiles, that's why I choose a kernel param as it could be added as part of the tuned profile. I thought about annotations but I'm a bit hesitant for two reasons:
- for security reasons, in k8s environments with untrusted users, giving the possibility to alter such behavior through pod annotations might not be ideal
- interactions between annotations and runtimes in k8s CRI as highlighted [here](https://github.com/kubernetes/cri-api/blob/30022f215cd7c176a48d567186d153476fc46f7b/pkg/apis/runtime/v1/api.proto#L462-L469).

Regarding CLI flag and environment variable, it would require additional configuration by using a dedicated wrapper to pass the CLI flag or modifying systemd unit for the environment variable, plus the environment variable could not be passed correctly to runtime by CRI or others, those options are not really convenient IMHO.

I found kernel param approach to be the less intrusive and less coupled

Reply:
I mentioned an annotation in config.json, not an annotation in Kubernetes pod
https://github.com/opencontainers/runtime-spec/blob/main/config.md#annotations

Reply:
Also curious why not just add a new proper JSON field to config.json ?

Reply:
Yes I understood, this fix needs to be enabled either system wide or per container primary in kubernetes environment running realtime applications without dealing with the OCI runtime spec, basically it should be deployed in environments with current tools like containerd/kubelet without any modifications at this level, so modifying the spec is not really an option. I mentionned about pod annotations as those can be passed to runtimes by CRI implementations like containerd/cri-o via OCI runtime spec annotations

Reply:
Pod annotations are not OCI annotations

Reply:
Right, I'm just talking about how it would work from a kubernetes pov to enable this feature by using an OCI annotation like you've suggested instead of using the kernel param.

AFAIK containerd allow to pass (un)filtered pod annotations for CRI container runtime via `pod_annotations`:

```
[plugins."io.containerd.grpc.v1.cri".containerd]
  default_runtime_name = "runc"
  [plugins."io.containerd.grpc.v1.cri".containerd.runtimes]
    [plugins."io.containerd.grpc.v1.cri".containerd.runtimes.runc]
      runtime_type = "io.containerd.runc.v2"
      base_runtime_spec = "/etc/containerd/cri-base.json"
      pod_annotations = ["*", "*/*"]
      container_annotations = ["*", "*/*"]
```

While looks like cri-o automatically pass pod annotations and store them as OCI annotations.

Such that a pod annotated with `org.opencontainers.runc.exec.isolated-cpu-affinity-transition` by example would enable the feature for runc for exec probes or `kubectl exec`.

Is it this approach that you would prefer ?

Reply:
@AkihiroSuda ^^

Reply:
> While looks like cri-o automatically pass pod annotations and store them as OCI annotations.

If this is true, it is a serious bug of cri-o.

> `org.opencontainers.runc.exec.isolated-cpu-affinity-transition`

SGTM

Reply:
how else would a k8s user specify annotations for the oci runtime?

Reply:
> how else would a k8s user specify annotations for the oci runtime? just image annotations??

No, non-admin user mustn't be able to specify OCI annotations that change the behavior of the runtime.

Reply:
@AkihiroSuda changes committed in https://github.com/opencontainers/runc/pull/3923/commits/197b60daec3f9257811f350da46f15adec24a8d4 to use OCI annotation instead of kernel param